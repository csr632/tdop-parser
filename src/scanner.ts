export interface Scanner {
  peek(): string | null;
  consume(): string | null;
}

export function createScanner(text: string): Scanner {
  const iterator = text[Symbol.iterator]();
  let next = iterator.next();
  let peeked = false;
  let token: string | null = null;

  return {
    peek(): string | null {
      if (!peeked) token = scanNextToken();
      peeked = true;
      return token;
    },
    consume(): string | null {
      if (peeked) {
        peeked = false;
        return token;
      }
      token = scanNextToken();
      return token;
    },
  };

  function scanNextToken(): string | null {
    // skip WhiteSpace at front
    while (!next.done && isWhiteSpace(next.value)) {
      next = iterator.next();
    }

    if (next.done) return null;

    if (isTokenBoundary(next.value)) {
      const token = next.value;
      next = iterator.next();
      return token;
    }

    let token = "";
    while (!next.done && !isTokenBoundary(next.value)) {
      token += next.value;
      next = iterator.next();
    }
    return token;
  }
}

// token is a sequence of \w
function isTokenBoundary(char: string) {
  if (!char) debugger;
  return !char.match(/^\w$/);
}

function isWhiteSpace(char: string) {
  return char.match(/\s/);
}
