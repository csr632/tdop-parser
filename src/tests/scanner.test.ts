import { expect, test } from "vitest";
import { createScanner } from "../scanner";

test("scanner", () => {
  const scanner = createScanner(`import { defineConfig } from 'vite';
  let test = 1234;
`);
  const tokens = [];
  while (scanner.peek()) {
    tokens.push(scanner.consume());
  }
  expect(tokens).toMatchSnapshot();
});
