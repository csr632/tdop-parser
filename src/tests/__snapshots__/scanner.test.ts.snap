// Vitest Snapshot v1

exports[`scanner 1`] = `
[
  "import",
  "{",
  "defineConfig",
  "}",
  "from",
  "'",
  "vite",
  "'",
  ";",
  "let",
  "test",
  "=",
  "1234",
  ";",
]
`;
